src/lib/mlang.js

```diff
@@ -147,108 +259,207 @@ foam.CLASS({
+// Add it back if it turns out we needed it.
+
+
+/** Binary expression for equality of two arguments. */
 foam.CLASS({
-  package: 'foam.mlang',
-  name: 'AndExpr',
-  implements: ['foam.mlang.Expr'],
-  properties: [
-    {
-      class: 'foam.mlang.ExprArray',
-      name: 'args'
+  package: 'foam.mlang.predicate',
+  name: 'EqExpr',
+  extends: 'foam.mlang.predicate.BinaryExpr',
+  methods: [
+    function f(o) {
+      return foam.util.equals(this.arg1.f(o), this.arg2.f(o));
     }
-  ],
+  ]
+});
+
+/** Binary expression for inequality of two arguments. */
+foam.CLASS({
+  package: 'foam.mlang.predicate',
+  name: 'NeqExpr',
+  extends: 'foam.mlang.predicate.BinaryExpr',
   methods: [
     function f(o) {
-      for ( var i = 0 ; i < this.args.length ; i++ ) {
-        if ( ! this.args[i].f(o) ) return false;
-      }
-      return true;
+      return ! foam.util.equals(this.arg1.f(o), this.arg2.f(o));
     }
   ]
 });
 
+/** Binary expression for "strictly less than". */
 foam.CLASS({
-  package: 'foam.mlang',
-  name: 'InExpr',
-  extends: 'foam.mlang.BINARY',
+  package: 'foam.mlang.predicate',
+  name: 'LtExpr',
+  extends: 'foam.mlang.predicate.BinaryExpr',
   methods: [
     function f(o) {
-      // TODO: This just looks like a reverse ContainsExpr
-      if ( this.arg2.f(o).indexOf(this.arg1.f(o)) != -1 ) return true;
-      return false;
+      return foam.util.compare(this.arg1.f(o), this.arg2.f(o)) < 0;
     }
   ]
 });
 
+/** Binary expression for "less than or equal to". */
 foam.CLASS({
-  package: 'foam.mlang',
-  name: 'EqExpr',
-  extends: 'foam.mlang.BINARY',
+  package: 'foam.mlang.predicate',
+  name: 'LteExpr',
+  extends: 'foam.mlang.predicate.BinaryExpr',
   methods: [
     function f(o) {
-      return foam.util.equals(this.arg1.f(o), this.arg2.f(o));
+      return foam.util.compare(this.arg1.f(o), this.arg2.f(o)) <= 0;
     }
   ]
 });
 
+/** Binary expression for "strictly greater than". */
 foam.CLASS({
-  package: 'foam.mlang',
+  package: 'foam.mlang.predicate',
   name: 'GtExpr',
-  extends: 'foam.mlang.BINARY',
+  extends: 'foam.mlang.predicate.BinaryExpr',
+  methods: [
+    function f(o) {
+      return foam.util.compare(this.arg1.f(o), this.arg2.f(o)) > 0;
+    }
+  ]
+});
+
+/** Binary expression for "greater than or equal to". */
+foam.CLASS({
+  package: 'foam.mlang.predicate',
+  name: 'GteExpr',
+  extends: 'foam.mlang.predicate.BinaryExpr',
   methods: [
     function f(o) {
-      return this.arg1.f(o) > this.arg2.f(o);
+      return foam.util.compare(this.arg1.f(o), this.arg2.f(o)) >= 0;
     }
   ]
 });
 
+
+/** Unary expression that checks the given property is well-defined. */
+foam.CLASS({
+  package: 'foam.mlang.predicate',
+  name: 'HasExpr',
+  extends: 'foam.mlang.predicate.UnaryExpr',
+
+  methods: [
+    function f(obj) {
+      var value = this.arg1.f(obj);
+      var notHas = value === undefined || value === null || value === '' ||
```

Comment:
Are these the right semantics?  Or should HAS be essentially hasOwnProperty()

If I want to query for things that have a property defined rather than things that are the defaultValue.  I'm not sure what is more useful or expected.